overlord/snapstate/handlers_components.go

```diff
@@ -0,0 +1,232 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snapdir"
+	"github.com/snapcore/snapd/timings"
+	"gopkg.in/tomb.v2"
+)
+
+// TaskComponentSetup returns the ComponentSetup and SnapSetup with task params hold
+// by or referred to by the task.
+func TaskComponentSetup(t *state.Task) (*ComponentSetup, *SnapSetup, error) {
+	snapsup, err := TaskSnapSetup(t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var compSetup ComponentSetup
+	err = t.Get("component-setup", &compSetup)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, nil, err
+	}
+	if err == nil {
+		return &compSetup, snapsup, nil
+	}
+
+	var id string
+	err = t.Get("component-setup-task", &id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ts := t.State().Task(id)
+	if ts == nil {
+		return nil, nil, fmt.Errorf("internal error: tasks are being pruned")
+	}
+	if err := ts.Get("component-setup", &compSetup); err != nil {
+		return nil, nil, err
+	}
+	return &compSetup, snapsup, nil
+}
+
+func (m *SnapManager) doPrepareComponent(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	compSetup, _, err := TaskComponentSetup(t)
+	if err != nil {
+		return err
+	}
+
+	if compSetup.Revision().Unset() {
+		// This is a local installation, revision is -1 (there
+		// is no history of local revisions for components).
+		compSetup.CompSideInfo.Revision = snap.R(-1)
+	}
+
+	t.Set("component-setup", compSetup)
+	return nil
+}
+
+func (m *SnapManager) doMountComponent(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	perfTimings := state.TimingsForTask(t)
+	compSetup, snapsup, err := TaskComponentSetup(t)
+	st.Unlock()
+	if err != nil {
+		return err
+	}
+
+	st.Lock()
+	deviceCtx, err := DeviceCtx(t.State(), t, nil)
+	st.Unlock()
+	if err != nil {
+		return err
+	}
+
+	// TODO we might want a checkComponents doing checks for some
+	// component types (see checkSnap and checkSnapCallbacks slice)
+
+	csi := compSetup.CompSideInfo
+	cpi := snap.MinimalComponentContainerPlaceInfo(csi.Component.ComponentName,
+		csi.Revision, snapsup.InstanceName(), snapsup.Revision())
+
+	cleanup := func() {
```

Comment:
I'm hesitant on this one as it required further modifications, for instance some code below should change to:
```
err = fmt.Errorf("expected component %q rev %v to be mounted but is not: %w",
			compSetup.CompSideInfo.Component, compSetup.Revision(), readInfoErr)
return err
```
The maintenance issue in this case is that future us need to make sure that usage of `err` is consistent with the check in the deferred function, which is a bit more subtle than direct calls to cleanup() that are easier to spot. On the other hand, calling `RemoveComponentDir()` is always safe as it will never actually error, but I still prefer to not call it on success.

Reply:
Great point, indeed naming the return makes this better. I've changed this now.